core/lib/multivm/src/versions/vm_latest/tracers/prestate_tracer.rs

```diff
@@ -0,0 +1,254 @@
+use std::{collections::HashMap, fmt, sync::Arc};
+
+use once_cell::sync::OnceCell;
+use zk_evm_1_4_1::tracing::{BeforeExecutionData, VmLocalStateData};
+use zksync_state::{StoragePtr, WriteStorage};
+use zksync_types::{
+    get_code_key, get_nonce_key, web3::signing::keccak256, AccountTreeId, Address, StorageKey,
+    StorageValue, H256, L2_ETH_TOKEN_ADDRESS, U256,
+};
+use zksync_utils::address_to_h256;
+
+use crate::{
+    interface::{dyn_tracers::vm_1_4_1::DynTracer, tracer::TracerExecutionStatus},
+    vm_latest::{
+        bootloader_state::BootloaderState,
+        old_vm::{history_recorder::HistoryMode, memory::SimpleMemory},
+        tracers::traits::VmTracer,
+        types::internals::ZkSyncVmState,
+    },
+};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Account {
+    pub balance: Option<U256>,
+    pub code: Option<U256>,
+    pub nonce: Option<U256>,
+    pub storage: Option<HashMap<H256, H256>>,
+}
+
+impl fmt::Display for Account {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "{{")?;
+        if let Some(balance) = self.balance {
+            writeln!(f, "  balance: \"0x{:x}\",", balance)?;
+        }
+        if let Some(code) = &self.code {
+            writeln!(f, "  code: \"{}\",", code)?;
+        }
+        if let Some(nonce) = self.nonce {
+            writeln!(f, "  nonce: {},", nonce)?;
+        }
+        if let Some(storage) = &self.storage {
+            writeln!(f, "  storage: {{")?;
+            for (key, value) in storage.iter() {
+                writeln!(f, "    {}: \"{}\",", key, value)?;
+            }
+            writeln!(f, "  }}")?;
+        }
+        writeln!(f, "}}")
+    }
+}
+
+type State = HashMap<Address, Account>;
+
+#[derive(Debug, Clone)]
+pub(crate) struct PrestateTracer {
+    pub pre: State,
+    pub post: State,
+    pub config: PrestateTracerConfig,
+    pub result: Arc<OnceCell<(State, State)>>,
+}
+
+impl PrestateTracer {
+    #[allow(dead_code)]
+    pub fn new(diff_mode: bool, result: Arc<OnceCell<(State, State)>>) -> Self {
+        Self {
+            pre: Default::default(),
+            post: Default::default(),
+            config: PrestateTracerConfig { diff_mode },
+            result,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct PrestateTracerConfig {
+    diff_mode: bool,
+}
+
+impl<S, H: HistoryMode> DynTracer<S, SimpleMemory<H>> for PrestateTracer {
+    fn before_execution(
+        &mut self,
+        _state: VmLocalStateData<'_>,
+        _data: BeforeExecutionData,
+        _memory: &SimpleMemory<H>,
+        _storage: StoragePtr<S>,
+    ) {
+    }
+}
+
+impl<S: WriteStorage, H: HistoryMode> VmTracer<S, H> for PrestateTracer {
+    fn initialize_tracer(&mut self, state: &mut ZkSyncVmState<S, H>) {
+        let initial_storage = state.storage.storage.inner().get_modified_storage_keys();
```

Comment:
This is empty at the beginning of the batch, what are you trying to achieve?

Reply:
Yes, you are right. When testing using the vmbuilder it was not empty at the beginning but the retrieval of the initial data of the storage should be on before_execution. Fixed